[FSDP2] Ran post-acc-grad hooks manually

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #129450

FSDP2 accumulates gradients for sharded parameters outside of the autograd engine's normal accumulation logic. We can respect registered post-accumulate-grad hooks by running them manually.

**Discussion**
Discussing with @soulitzer, changing FSDP2 to make the sharded parameters autograd leaves requires nontrivial changes to FSDP and some changes to the autograd engine (around forward vs. backward streams) where the changes may not preserve eager-mode performance and/or add some complexity.

Under the FSDP2 design, the sharded parameters never participate in autograd, so calling `register_post_accumulate_grad_hook` on them would otherwise be a no-op. In other words, there is virtually no chance for FSDP2 incorrectly re-running the hook when it should not.

Given these, a reasonable near-term solution is for FSDP2 to run the post-accumulate-grad hooks manually.

**Caveats**
- Running `foreach=False` optimizer _per parameter tensor_  incurs significantly higher CPU overhead compared to `foreach=True` (partially due to `DTensor` being a `__torch_dispatch__` tensor subclass).
    - On preliminary benchmarking on Llama3-8B on 8 GPUs, this CPU overhead is mostly tolerable, but on smaller # of GPUs or a less compute-intensive model, this may not be.
    - One solution for native Adam/AdamW is to use `fused=True`, which makes both the CPU overhead lower and GPU compute faster. However, this is generally not an option for user-defined optimizers.
    - If this CPU overhead blocks adoption of this feature, then we should seriously consider an FSDP-specific API like `register_post_backward_hook(params: List[nn.Parameter]) -> None` that allows the user to see all parameters in the `FSDPParamGroup` together for the hook so that the user can still run a `foreach=True` optimizer step on that `List[nn.Parameter]`.
- The post-accumulate-grad hook runs in the reduce-scatter stream. Our current stream handling logic does not have the default stream wait for the reduce-scatter stream until the end of backward. Unless we add that, we cannot simply run the post-accumulate-grad hook in the default stream.
    - This means that optimizer compute will overlap with backward compute, which may slowdown end-to-end execution slightly (e.g. due to SM contention or wave quantization effects). For example, on Llama3-8B, we see about ~3% decrease in MFU when running optimizer in backward even though the optimizer steps are fully overlapped and there are no CPU boundedness issues.
- This PR's goal is only to run the hook manually. State dict etc. for optimizer-in-backward is out of scope.

**Experiments (torchtitan)**
- Llama3-8B on 2 GPUs, local batch size 1, with full activation checkpointing, and bf16/fp32 mixed precision:
    - Without optimizer-in-backward: 82.03 GiB reserved memory; 28.1% MFU
    - With optimizer-in-backward (`foreach=False`): 72.84 GiB reserved memory; 28.9% MFU (speedup from more of optimizer step overlapped)
    - With optimizer-in-backward (`fused=True`): 70.84 GiB reserved memory; 30.4% MFU


cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k

Commit:
[FSDP2] Ran post-acc-grad hooks manually

[ghstack-poisoned]

awgu commented on torch/distributed/_composable/fsdp/_fsdp_collectives.py:
```diff
@@ -336,6 +336,11 @@ def foreach_reduce(
                 )
                 fsdp_param.sharded_param.grad = new_sharded_dtensor_grad
+            for hook in (
+                getattr(fsdp_param.sharded_param, "_post_accumulate_grad_hooks", {})
```
Currently, `torch.Tensor`s should always have `_post_accumulate_grad_hooks` defined, regardless of if leaf or if a hook is actually registered, so the `getattr` is not necessary. This is gray area anyway since we are relying on a private attribute though.

weifengpy replied:
totally impressed by the minimal lines of change to showcase optimizer-in-the-backward

awgu commented:
@pytorchbot merge

clee2000 commented:
@pytorchbot revert -m "broke distributed/_composable/fsdp/test_fully_shard_compile.py::TestFullyShardCompile::test_simple_mlp_fullgraph_backend_aot_eager https://hud.pytorch.org/pytorch/pytorch/commit/7ebffef4d02a3cc68dbbcf44b92d63c7fe0ebb67 https://github.com/pytorch/pytorch/actions/runs/9667812641/job/26671489454.  Test got added in https://github.com/pytorch/pytorch/pull/129157 which is before your mergebase" -c landrace

awgu commented on torch/distributed/_composable/fsdp/_fsdp_collectives.py:
```diff
@@ -336,6 +336,11 @@ def foreach_reduce(
+            for hook in (
+                getattr(fsdp_param.sharded_param, "_post_accumulate_grad_hooks", {})
+                or {}
+            ).values():
```
@yf225 What should we do here? Dynamo does not support this `getattr` call.

```
File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py", line 1480, in CALL_FUNCTION
2024-06-25T20:11:10.8946722Z     self.call_function(fn, args, {})
2024-06-25T20:11:10.8947571Z   File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py", line 754, in call_function
2024-06-25T20:11:10.8947838Z     self.push(fn.call_function(self, args, kwargs))
2024-06-25T20:11:10.8948692Z   File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/_dynamo/variables/builtin.py", line 963, in call_function
2024-06-25T20:11:10.8948882Z     return handler(tx, args, kwargs)
2024-06-25T20:11:10.8949707Z   File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/_dynamo/variables/builtin.py", line 712, in <lambda>
2024-06-25T20:11:10.8949968Z     return lambda tx, args, kwargs: obj.call_function(
2024-06-25T20:11:10.8950846Z   File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/_dynamo/variables/builtin.py", line 963, in call_function
2024-06-25T20:11:10.8951019Z     return handler(tx, args, kwargs)
2024-06-25T20:11:10.8951892Z   File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/_dynamo/variables/builtin.py", line 847, in builtin_dipatch
2024-06-25T20:11:10.8952082Z     rv = fn(tx, args, kwargs)
2024-06-25T20:11:10.8952978Z   File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/_dynamo/variables/builtin.py", line 765, in call_self_handler
2024-06-25T20:11:10.8953209Z     result = self_handler(tx, *args, **kwargs)
2024-06-25T20:11:10.8954068Z   File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/_dynamo/variables/builtin.py", line 1564, in call_getattr
2024-06-25T20:11:10.8954323Z     hasattr_var = self.call_hasattr(tx, obj, name_var)
2024-06-25T20:11:10.8955350Z   File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/_dynamo/variables/builtin.py", line 1471, in call_hasattr
2024-06-25T20:11:10.8955649Z     return obj.call_hasattr(tx, name)
2024-06-25T20:11:10.8956490Z   File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/_dynamo/variables/base.py", line 296, in call_hasattr
2024-06-25T20:11:10.8956802Z     unimplemented(f"hasattr {self.__class__.__name__} {name}")
2024-06-25T20:11:10.8957567Z   File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/_dynamo/exc.py", line 221, in unimplemented
2024-06-25T20:11:10.8957753Z     raise Unsupported(msg)
2024-06-25T20:11:10.8958218Z torch._dynamo.exc.Unsupported: hasattr TensorVariable _post_accumulate_grad_hooks
```

yf225 replied:
@awgu I believe there is a way to make this work - for now would you like to skip this path under compile? I can come back to make this work for compile later

awgu replied:
Yep skipping this for now is good.

Commit:
Update on "[FSDP2] Ran post-acc-grad hooks manually"


FSDP2 accumulates gradients for sharded parameters outside of the autograd engine's normal accumulation logic. We can respect registered post-accumulate-grad hooks by running them manually.

**Discussion**
Discussing with soulitzer, changing FSDP2 to make the sharded parameters autograd leaves requires nontrivial changes to FSDP and some changes to the autograd engine (around forward vs. backward streams) where the changes may not preserve eager-mode performance and/or add some complexity.

Under the FSDP2 design, the sharded parameters never participate in autograd, so calling `register_post_accumulate_grad_hook` on them would otherwise be a no-op. In other words, there is virtually no chance for FSDP2 incorrectly re-running the hook when it should not.

Given these, a reasonable near-term solution is for FSDP2 to run the post-accumulate-grad hooks manually.

**Caveats**
- Running `foreach=False` optimizer _per parameter tensor_  incurs significantly higher CPU overhead compared to `foreach=True` (partially due to `DTensor` being a `__torch_dispatch__` tensor subclass).
    - On preliminary benchmarking on Llama3-8B on 8 GPUs, this CPU overhead is mostly tolerable, but on smaller # of GPUs or a less compute-intensive model, this may not be.
    - One solution for native Adam/AdamW is to use `fused=True`, which makes both the CPU overhead lower and GPU compute faster. However, this is generally not an option for user-defined optimizers.
    - If this CPU overhead blocks adoption of this feature, then we should seriously consider an FSDP-specific API like `register_post_backward_hook(params: List[nn.Parameter]) -> None` that allows the user to see all parameters in the `FSDPParamGroup` together for the hook so that the user can still run a `foreach=True` optimizer step on that `List[nn.Parameter]`.
- The post-accumulate-grad hook runs in the reduce-scatter stream. Our current stream handling logic does not have the default stream wait for the reduce-scatter stream until the end of backward. Unless we add that, we cannot simply run the post-accumulate-grad hook in the default stream.
    - This means that optimizer compute will overlap with backward compute, which may slowdown end-to-end execution slightly (e.g. due to SM contention or wave quantization effects). For example, on Llama3-8B, we see about ~3% decrease in MFU when running optimizer in backward even though the optimizer steps are fully overlapped and there are no CPU boundedness issues.
- This PR's goal is only to run the hook manually. State dict etc. for optimizer-in-backward is out of scope.

**Experiments (torchtitan)**
- Llama3-8B on 2 GPUs, local batch size 1, with full activation checkpointing, and bf16/fp32 mixed precision:
    - Without optimizer-in-backward: 82.03 GiB reserved memory; 28.1% MFU
    - With optimizer-in-backward (`foreach=False`): 72.84 GiB reserved memory; 28.9% MFU (speedup from more of optimizer step overlapped)
    - With optimizer-in-backward (`fused=True`): 70.84 GiB reserved memory; 30.4% MFU


cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

awgu commented on torch/distributed/_composable/fsdp/_fsdp_collectives.py:
```diff
@@ -351,6 +351,12 @@ def foreach_reduce(
                     new_sharded_grad
                 )
                 fsdp_param.sharded_param.grad = new_sharded_dtensor_grad
+            if not ca.compiled_autograd_enabled:
```
Change for re-land is here. We skip running these hooks if tracing with compiled autograd.

awgu commented:
@pytorchbot merge

awgu commented:
@pytorchbot rebase -s

Commit:
Update

[ghstack-poisoned]

awgu commented:
@pytorchbot merge

awgu commented:
[pull / linux-focal-cuda12.1-py3.10-gcc9-sm86 / test (default, 4, 5, linux.g5.4xlarge.nvidia.gpu)](https://github.com/pytorch/pytorch/actions/runs/9704872494/job/26786706175) failures look unrelated:
```
The following tests failed consistently: ['test/test_foreach.py::TestForeachCUDA::test_unary_op_tensors_on_different_devices__foreach_abs_cuda_complex64']
```

awgu commented:
@pytorchbot merge -i

awgu commented:
@pytorchbot merge

awgu commented:
<img width="850" alt="Screenshot 2024-06-28 at 10 42 26 AM" src="https://github.com/pytorch/pytorch/assets/31054793/dcb82fdc-2b13-422f-aaa4-870a717fbf42">
merge bot seems confused -- llm retrieval tests were marked unstable already and can be ignored

awgu commented:
@pytorchbot merge -i
